src/pepr/operator/controllers/istio/namespace.ts

```diff
@@ -0,0 +1,315 @@
+  if (pkg.spec?.network?.serviceMesh?.mode === Mode.Ambient) {
+    // Check if ambient mode is available
+    if (!UDSConfig.isAmbientDeployed) {
+      // Ambient mode requested but not available, fall back to sidecar mode
+      log.warn(
+        `Ambient mode requested for package ${pkg.metadata.name} but Ambient is not deployed. Falling back to sidecar mode.`,
+      );
+      await writeEvent(pkg, {
+        message:
+          "Ambient mode requested but Ambient is not deployed. Falling back to sidecar mode.",
+        reason: "AmbientUnavailable",
+        type: "Warning",
+      });
+
+      // Fall back to sidecar mode
+      targetIstioState = IstioState.Sidecar;
+    } else {
+      // Ambient mode is available and requested
+      targetIstioState = IstioState.Ambient;
+    }
+  } else {
+    // Sidecar mode requested/by default
+    targetIstioState = IstioState.Sidecar;
+  }
+
+  const result = getIstioLabels(labels, targetIstioState, currentIstioState);
+
+  // Apply namespace updates and restart pods if needed
+  await applyNamespaceUpdates(
+    pkg.metadata.namespace,
+    result.labels,
+    annotations,
+    sourceNS.metadata?.labels,
+    sourceNS.metadata?.annotations,
+  );
+
+  await restartPodsIfNeeded(pkg.metadata.namespace, result.shouldRestartPods, targetIstioState);
+}
+
+/**
+ * Cleanup the namespace by removing Istio labels if there are no remaining packages
+ * Or adjust the Istio mode if existing packages change it
+ *
+ * @param pkg the package to cleanup
+ */
+export async function cleanupNamespace(pkg: UDSPackage) {
+  if (!pkg.metadata?.namespace || !pkg.metadata.name) {
+    throw new Error(`Invalid Package definition, missing namespace or name`);
+  }
+
+  const sourceNS = await K8s(kind.Namespace).Get(pkg.metadata.namespace);
+  const labels = { ...(sourceNS.metadata?.labels || {}) };
+  const annotations = { ...(sourceNS.metadata?.annotations || {}) };
+  const currentState = getCurrentIstioState(labels);
+  const originalIstioState = annotations[ISTIO_STATE_ANNOTATION] as IstioState;
+
+  // Remove the package annotation
+  delete annotations[`uds.dev/pkg-${pkg.metadata.name}`];
+
+  // Check if there are any other package annotations
+  // Backwards compatibility for multiple package CRs in a single namespace
+  const hasOtherPackages = Object.keys(annotations).some(key => key.startsWith("uds.dev/pkg-"));
+
+  // Only modify Istio labels if this is the last package
+  let result;
+  if (hasOtherPackages) {
+    // Keep existing labels if other packages are still present, don't cycle pods/change istio state
+    result = { labels, shouldRestartPods: false };
+  } else {
+    // Set labels based on the original state and determine if pods need to be restarted
+    result = getIstioLabels(labels, originalIstioState, currentState);
+
+    // Delete the annotation since we're restoring to original state
+    delete annotations[ISTIO_STATE_ANNOTATION];
+  }
```

Comment:
Highlighting this behavior - I think we want this to ensure that we don't break cleanup of packages when multiple are present. This logic would only remove the pkg annotation from the namespace if other packages are still present, not performing any label changes or pod restarts. This allows removing a second package, but does not support any complex state changes since we don't plan to support multiple package CRs in a single namespace going forward.